### PR TITLE
feat(golang): array support and binary protocol for doubles

### PR DESCRIPTION
**Pull request by kafka1991**
1. array support
2. protocol version option, binary protocol for doubles

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -99,6 +100,95 @@ HTTP is the recommended transport to use. To connect via TCP, set the configurat
...
+    Symbol("dataset", "training_batch_1").
+    Float64NDArrayColumn("features", arr).
+    AtNow(ctx)
+```
+
+The array data is sent over a new protocol version (2) that is auto-negotiated
+when using HTTP(s), or can be specified explicitly via the ``protocol_version=2``
+parameter when using TCP(s).
+
+We recommend using HTTP(s), but here is an TCP example, should you need it::
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply README.md]: fixed in https://github.com/questdb/go-questdb-client/pull/54/commits/9d702b020cc8ea515e1c17a80a0dcfd2251bdbe5

File: README.md [hunk 2]
```diff
@@ -99,6 +100,95 @@ HTTP is the recommended transport to use. To connect via TCP, set the configurat
...
+
+### 1D Arrays
+
+```go
+// Send a 1D array of doubles
+values1D := []float64{1.1, 2.2, 3.3, 4.4}
+err = sender.
+    Table("measurements").
+    Symbol("sensor", "temp_probe_1").
+    Float641DArrayColumn("readings", values1D).
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply README.md]: renamed to `Float64Array1DColumn` in https://github.com/questdb/go-questdb-client/pull/54/commits/9d702b020cc8ea515e1c17a80a0dcfd2251bdbe5

To avoid introducing a breaking change when we support more types in the future, we didn't use `Array1DColumn` here.

File: buffer.go [hunk 3]
```diff
@@ -26,19 +26,67 @@ package questdb
...
 
+type binaryFlag byte
+
+const (
+	arrayBinaryFlag   binaryFlag = 14
+	float64BinaryFlag binaryFlag = 16
+)
+
+// isLittleEndian checks if the current machine uses little-endian byte order
+func isLittleEndian() bool {
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply buffer.go]: fixed in https://github.com/questdb/go-questdb-client/commit/9d702b020cc8ea515e1c17a80a0dcfd2251bdbe5

File: conf_parse.go [hunk 4]
```diff
@@ -162,6 +162,18 @@ func confFromStr(conf string) (*lineSenderConfig, error) {
...
 			return nil, NewInvalidConfigStrError("tls_roots_password is not available in the go client")
+		case "protocol_version":
+			if v != "auto" {
+				version, err := strconv.Atoi(v)
+				if err != nil {
+					return nil, NewInvalidConfigStrError("invalid %s value, %q is not a valid int", k, v)
+				}
+				pVersion := protocolVersion(version)
+				if pVersion < ProtocolVersion1 || pVersion > ProtocolVersion2 {
+					return nil, NewInvalidConfigStrError("current client only supports protocol version 1(text format for all datatypes), 2(binary format for part datatypes) or explicitly unset")
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply conf_parse.go]: fixed in https://github.com/questdb/go-questdb-client/commit/9d702b020cc8ea515e1c17a80a0dcfd2251bdbe5

File: examples/from-conf/main.go [hunk 5]
```diff
@@ -27,12 +27,28 @@ func main() {
...
+	// Prepare array data.
+	// QuestDB server version 9.0.0 or later is required for array support.
+	array, err := qdb.NewNDArray[float64](2, 3, 2)
+	if err != nil {
+		log.Fatal(err)
+	}
+	hasMore := true
+	val := 100.0
+	for hasMore {
+		hasMore, err = array.Append(val + 1)
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply examples/from-conf/main.go]: Here I want to highlight the good practices for ndarray reuse.

File: http_sender.go [hunk 6]
```diff
@@ -399,6 +440,90 @@ func (s *httpLineSender) makeRequest(ctx context.Context) (bool, error) {
 
 }
 
+func detectProtocolVersion(ctx context.Context, conf *lineSenderConfig) (protocolVersion, error) {
+	tmpClient := http.Client{
+		Transport: globalTransport.transport,
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply http_sender.go]: fixed in https://github.com/questdb/go-questdb-client/commit/9d702b020cc8ea515e1c17a80a0dcfd2251bdbe5

File: http_sender.go [hunk 7]
```diff
@@ -399,6 +440,90 @@ func (s *httpLineSender) makeRequest(ctx context.Context) (bool, error) {
...
+
+	switch resp.StatusCode {
+	case 404:
+		return ProtocolVersion1, nil
+	case 200:
+		return parseServerSettings(resp, conf)
+	default:
+		buf, _ := io.ReadAll(resp.Body)
+		return protocolVersionUnset, fmt.Errorf("failed to detect server line protocol version [http-status=%d, http-message=%s]",
+			resp.StatusCode, string(buf))
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply http_sender.go:475]: fixed in https://github.com/questdb/go-questdb-client/commit/9d702b020cc8ea515e1c17a80a0dcfd2251bdbe5

File: integration_test.go [hunk 8]
```diff
@@ -132,7 +132,7 @@ func setupQuestDB0(ctx context.Context, auth ilpAuthType, setupProxy bool) (*que
 		return nil, err
 	}
 	req := testcontainers.ContainerRequest{
-		Image:          "questdb/questdb:7.4.2",
+		Image:          "questdb/questdb:9.0.1",
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply integration_test.go]: fixed in https://github.com/questdb/go-questdb-client/commit/9d702b020cc8ea515e1c17a80a0dcfd2251bdbe5

File: integration_test.go [hunk 9]
```diff
@@ -413,44 +413,103 @@ func (suite *integrationTestSuite) TestE2EValidWrites() {
 				Count: 1,
 			},
 		},
+		{
+			"double array",
+			testTable,
+			func(s qdb.LineSender) error {
+				values1D := []float64{1.0, 2.0, 3.0, 4.0, 5.0}
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply integration_test.go:420]: Tests added in https://github.com/questdb/go-questdb-client/commit/9d702b020cc8ea515e1c17a80a0dcfd2251bdbe5

File: buffer.go [hunk 10]
```diff
@@ -517,6 +571,188 @@ func (b *buffer) Float64Column(name string, val float64) *buffer {
...
+	// Write values
+	if len(values) > 0 {
+		b.writeFloat64Data(values)
+	}
+
+	b.hasFields = true
+	return b
+}
+
+func (b *buffer) Float642DArrayColumn(name string, values [][]float64) *buffer {
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply buffer.go]: Yes, they also need to validate. fixed in https://github.com/questdb/go-questdb-client/commit/9d702b020cc8ea515e1c17a80a0dcfd2251bdbe5

File: ndarray.go [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+func (n *NdArray[T]) Append(val T) (bool, error) {
+	if n.appendIndex >= uint(len(n.data)) {
+		return false, errors.New("array is full")
+	}
+	n.data[n.appendIndex] = val
+	n.appendIndex++
+	return n.appendIndex < uint(len(n.data)), nil
+}
+
+// ResetAppendIndex resets the append index to 0
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply ndarray.go]: Optimize godoc in https://github.com/questdb/go-questdb-client/commit/9d702b020cc8ea515e1c17a80a0dcfd2251bdbe5

File: buffer.go [hunk 12]
```diff
@@ -517,6 +571,188 @@ func (b *buffer) Float64Column(name string, val float64) *buffer {
...
+	var dim2, dim3 int
+	if dim1 > 0 {
+		dim2 = len(values[0])
+		if dim2 > 0 {
+			dim3 = len(values[0][0])
+		}
+
+		for i, level1 := range values {
+			if len(level1) != dim2 {
+				b.lastErr = fmt.Errorf("irregular 3D array shape: level1[%d] has length %d, expected %d", i, len(level1), dim2)
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply buffer.go:710]: Tests add in https://github.com/questdb/go-questdb-client/commit/9d702b020cc8ea515e1c17a80a0dcfd2251bdbe5

File: ndarray.go [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+
+	// Calculate flat index
+	index := 0
+	for i, pos := range positions {
+		index += int(pos) * int(product(n.shape[i+1:]))
+	}
+	return index, nil
+}
+
+// validateShape validates that shape dimensions are valid
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply ndarray.go]: Removed in https://github.com/questdb/go-questdb-client/commit/9d702b020cc8ea515e1c17a80a0dcfd2251bdbe5

File: http_sender.go [hunk 14]
```diff
@@ -399,6 +440,90 @@ func (s *httpLineSender) makeRequest(ctx context.Context) (bool, error) {
...
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case 404:
+		return ProtocolVersion1, nil
+	case 200:
+		return parseServerSettings(resp, conf)
+	default:
+		buf, _ := io.ReadAll(resp.Body)
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply http_sender.go:470]: Tests added in https://github.com/questdb/go-questdb-client/commit/9d702b020cc8ea515e1c17a80a0dcfd2251bdbe5

**Comment by kafka1991** [comment]
@puzpuzpuz Thanks for your review!

File: integration_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply integration_test.go:420]: Could you also add a test case with null array elements, i.e. some elements that have NaN values? Those elements should be interpreted as nulls by server.

**Comment by puzpuzpuz** [comment]
> Do you know how should we trigger the CI check?

I've asked if someone can configure that on this repo - I don't have the permissions.

File: buffer.go [hunk 15]
```diff
@@ -517,6 +573,206 @@ func (b *buffer) Float64Column(name string, val float64) *buffer {
...
+		return b
+	}
+
+	// Validate array shape
+	dim1 := len(values)
+	var dim2 int
+	if dim1 > 0 {
+		dim2 = len(values[0])
+		totalElements := dim1 * dim2
+		if dim1 > MaxArrayElements || dim2 > MaxArrayElements || totalElements > MaxArrayElements || totalElements < 0 {
```
Comment (puzpuzpuz) [inline buffer.go:651]: We need to improve overflow detection:
dim1 = 65536, dim2 = 65536
totalElements = 65536 * 65536 = 4294967296
On 32-bit systems: overflows to 0, bypasses validation

File: ndarray.go [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+		return errors.New("shape cannot be empty")
+	}
+
+	if len(shape) > MaxDimensions {
+		return fmt.Errorf("too many dimensions: %d exceeds maximum of %d",
+			len(shape), MaxDimensions)
+	}
+
+	totalElements := product(shape)
+	if totalElements > MaxArrayElements {
```
[the remark this answers is not in the document]
Comment (kafka1991) [reply ndarray.go:201]: fixed in https://github.com/questdb/go-questdb-client/pull/54/commits/99d1a6c94e8bfe8d5bad0c055c7f13d8e6dbd2e0

**Comment by puzpuzpuz** [comment]
@kafka1991 Vlad copied settings from the core repo to this one, but I still don't see the Approve and run CI button. Maybe a new commit needs to be pushed into your branch to have GH reevaluate the settings?

**Comment by kafka1991** [comment]
Continue in a separate PR #55
